credoai/lens/lens.py

```diff
@@ -172,9 +179,16 @@ def run(self):
         """
         if self.pipeline == []:
             raise RuntimeError("No evaluators were added to the pipeline.")
-        for step in self.pipeline:
-            self.logger.info(f"Running evaluation for step: {step}")
-            step.evaluator.evaluate()
+
+        # Run evaluators in parallel. Shared object (self.pipeline) necessitates writing
+        # results to intermediate object evaluator_results
+        evaluator_results = Parallel(n_jobs=self.n_jobs, verbose=100)(
+            delayed(step.evaluator.evaluate)() for step in self.pipeline
+        )
+
+        # Write intermediate evaluator results back into self.pipeline for later processing
+        for idx, evaluator in enumerate(evaluator_results):
+            self.pipeline[idx].evaluator = evaluator
```

Comment:
I don't know exactly how `joblib` works, but this assumes that `evaluator_results` is always in the same order as the pipeline steps. Is that always the case?

If the second in the pipeline finishes earlier than the first, does joblib waits to populate evaluator_results?

Reply:
So I had tested this a little bit before pushing, but now I've done some more research to verify.

1. Wrote a function "delay_sqrt(i)" which calls time.sleep(2) if i is odd (no sleep if even) and tested --> results in order
2. Added a sleep(10) call to some of the evaluators --> results still in order
3. Per [this](https://stackoverflow.com/questions/56659294/does-joblib-parallel-keep-the-original-order-of-data-passed), multiprocessing (which I'm not using) guarantees that results are in order...I don't see that explicitly in the joblib docs. Going to [multiprocessing.Pool](https://docs.python.org/3.5/library/multiprocessing.html#multiprocessing.pool.Pool.map), the map function is what's being called and map guarantees return order, but not execution order (which is fine, but maybe not speed optimal)
4. Loky backend (which we are using) -> [implements a map function ](https://github.com/joblib/joblib/externals/loky/process_executor.py) based on the multiprocessing map function linked above (so parallel execution, finish order not guaranteed, return order is guaranteed). If you click on that last link to the loky definition, you can see that it's actually straight up calling the concurrent.futures map, which is based on vanilla Python [map](https://www.geeksforgeeks.org/python-map-function/). My interpretation of that is that under the hood what happens is i) lay out a result vector whose members are directly linked to function + argument pairs, ii) spin out the function + argument pairs to run in parallel using different processes, iii) as they return, the process is placed in a specific spot in the result list via the PID of that process (see [this](https://github.com/joblib/joblib/blob/master/joblib/externals/loky/process_executor.py#L670) function), and then iv) when everything is done all the processes clean-up and you have your in-order list according to order of arguments, not necessarily order of execution.

Reply:
Sounds great! Excellent documentation!